### PR TITLE
Sankey 3.2.2.0v

### PR DESCRIPTION
**Pull request by kullJul**

### CODE REVIEW

File: src/sankeyDiagram.ts [hunk 1]
```diff
@@ -214,6 +216,9 @@ export class SankeyDiagram implements IVisual {
         values: "Weight"
     };
 
+    public static InputTooltipValue: string = "Input ";
+    public static OutputTooltipValue: string = "Output ";
```
Comment (v-ruboranov) [inline src/sankeyDiagram.ts:220]: Why do we need space in the end of the string?

File: src/sankeyDiagram.ts [hunk 2]
```diff
@@ -230,6 +235,7 @@ export class SankeyDiagram implements IVisual {
     private curvatureOfLinks: number = 0.5;
 
     private static NodeAndBackwardLinkDistance = 5;
+    private static DistanceBetweenLinks = 3;
```
Comment (v-ruboranov) [inline src/sankeyDiagram.ts:238]: Type of _**DistanceBetweenLinks**_  variable is absent

File: src/sankeyDiagram.ts [hunk 3]
```diff
@@ -911,29 +897,42 @@ export class SankeyDiagram implements IVisual {
...
                 value: nodeName
-            }, {
-                displayName: valueDisplayName || SankeyDiagram.RoleNames.values,
-                value: formattedNodeWeight
             }
         ];
+
+        if (valueDisplayName) {
+            tooltips.push({
+                displayName: SankeyDiagram.InputTooltipValue + valueDisplayName || SankeyDiagram.RoleNames.values,
```
Comment (v-ruboranov) [inline src/sankeyDiagram.ts:926]: is '+' a good idea of string concatenation?

File: src/sankeyDiagram.ts [hunk 4]
```diff
@@ -2038,98 +2036,116 @@ export class SankeyDiagram implements IVisual {
...
 
-    private getSvgPathForBackwardLink(link: SankeyDiagramLink) {
+    private getSvgPathForBackwardLink(link: SankeyDiagramLink, minHeight: number) {
         let pathParams: string = "";
-        const distanceBetweenLinks: number = 3;
-        const distanceFromNodeToLinks: number = 5;
 
-        const fixedLinkHeight = link.height - distanceBetweenLinks;
-        const linkKneeSize: number = link.height;
+        const fixedLinkHeight = link.height - SankeyDiagram.DistanceBetweenLinks;
```
Comment (v-ruboranov) [inline src/sankeyDiagram.ts:2042]: type of **_fixedLinkHeight_** variable is absent

File: src/sankeyDiagram.ts [hunk 5]
```diff
@@ -2038,98 +2036,116 @@ export class SankeyDiagram implements IVisual {
...
+        curveRadius = (link.dyDestination + SankeyDiagram.NodeAndBackwardLinkDistance)/2; 
+        curveCenterY = link.destination.y - SankeyDiagram.NodeAndBackwardLinkDistance + curveRadius;
+       
+        pathParams +=
+            `C ${curveCenterX - link.destination.width/2} ${link.destination.y + link.dyDestination + SankeyDiagram.DistanceBetweenLinks/2},` +
+            ` ${curveCenterX + linkInnerKneeSize} ${link.destination.y + link.dyDestination + SankeyDiagram.DistanceBetweenLinks/2} ,` +
+            ` ${curveCenterX + linkInnerKneeSize} ${curveCenterY}`;      
 
-        pathParams += `L ${link.destination.x + link.destination.width + distanceFromNodeToLinks} ${link.destination.y - SankeyDiagram.NodeAndBackwardLinkDistance}`;
+        pathParams +=
```
Comment (v-ruboranov) [inline src/sankeyDiagram.ts:2100]: Is _**pathParams**_ variable's several additions can be separated to other function or calculated at the end?

**Comment by kullJul** [comment]
> LGTM, a lot of calc changed, test pass?
yes, adding a link to the build job and a screenshot from running tests locally
https://github.com/microsoft/powerbi-visuals-sankey/actions/runs/6034458822/job/16372866878?pr=161
image

File: src/sankeyDiagram.ts (hunk 2 above)
Comment (kullJul) [reply src/sankeyDiagram.ts]: fixed 
https://github.com/microsoft/powerbi-visuals-sankey/pull/161/commits/0f90ad9016883a0d20c0bf7a137e62b2f99357cf

File: src/sankeyDiagram.ts (hunk 4 above)
Comment (kullJul) [reply src/sankeyDiagram.ts]: fixed in this commit https://github.com/microsoft/powerbi-visuals-sankey/pull/161/commits/0f90ad9016883a0d20c0bf7a137e62b2f99357cf

File: style/visual.less [hunk 6]
```diff
@@ -44,6 +44,16 @@
                 shape-rendering: crispEdges;
             }
 
+            .nodeRect:focus {
+                outline: none;
+            }
+
+            .nodeRect:focus-visible {
```
Comment (MulyukovAidar) [inline style/visual.less:51]: works on Edge, Chrome, Firefox?

File: style/visual.less [hunk 7]
```diff
@@ -68,14 +78,15 @@
...
-    }
 
-    &:focus {
-        outline: none;
-    }
-    
-    &:focus-visible {
-        outline: auto 1px;
-        outline-color: -webkit-focus-ring-color;
+        .link:focus {
```
Comment (MulyukovAidar) [inline style/visual.less:82]: try using _less_ hierarchy like here
https://riptutorial.com/less/example/29941/pseudo-elements

**Review by MulyukovAidar — CHANGES_REQUESTED** [review]

File: style/visual.less (hunk 7 above)
Comment (kullJul) [reply style/visual.less]: fixed a68bc7a8f90c346f0e47de506aa77473929b70e3

File: style/visual.less (hunk 6 above)
Comment (kullJul) [reply style/visual.less]: works as expected on Edge, Chrome and Firefox

File: src/sankeyDiagram.ts [hunk 8]
```diff
@@ -2038,98 +2036,115 @@ export class SankeyDiagram implements IVisual {
...
-
-        if (link.destination.x < link.source.x) {
-            x0 = link.source.x;
-            x1 = link.destination.x + link.destination.width;
-        } else {
-            x0 = link.source.x + link.source.width;
-            x1 = link.destination.x;
-        }
+            y1: number,
+            curveRadius: number,
```
[the remark this answers is not in the document]
Comment (kullJul) [reply src/sankeyDiagram.ts:2025]: fixed 9641777a4a34e067384b49a37a762f9b7b33b5f6

File: src/sankeyDiagram.ts [hunk 9]
```diff
@@ -1523,6 +1517,10 @@ export class SankeyDiagram implements IVisual {
                         .darker(SankeyDiagram.StrokeColorFactor)
                         .toString()
             )
+            .attr("tabindex", 0)
```
[the remark this answers is not in the document]
Comment (kullJul) [reply src/sankeyDiagram.ts]: fixed b7233833b51c708ac1df35884fafa57704f8481d

File: src/sankeyDiagram.ts [hunk 10]
```diff
@@ -1434,16 +1434,10 @@ export class SankeyDiagram implements IVisual {
                 }
 
                 if (link.source === node) {
-                    if (link.direction === SankeyLinkDirrections.Forward)
-                        link.dySource = shiftByAxisY;
-                    else
-                        link.dyDestination = shiftByAxisY;
+                    link.dySource = shiftByAxisY;
```
[the remark this answers is not in the document]
Comment (kullJul) [reply src/sankeyDiagram.ts]: renamed 508dec9d6135b9663334121c77b64a0dd8b6a368

File: src/sankeyDiagram.ts (hunk 3 above)
Comment (kullJul) [reply src/sankeyDiagram.ts]: fixed and localized 09b4985a6c4338db9d540ec2b4cc86a342279eea

File: test/visualTest.ts [hunk 11]
```diff
@@ -804,117 +804,164 @@ describe("SankeyDiagram", () => {
...
         });
 
         it("enter toggles the correct slice", (done: DoneFn) => {
             const enterEvent = new KeyboardEvent("keydown", { code: "Enter", bubbles: true });
+            visualBuilder.updateRenderTimeout(
+                dataView, () => {
+                        const links: NodeListOf<HTMLElement> = visualBuilder.linkElements;
+
+                        links[0].dispatchEvent(enterEvent);
+
```
Comment (MulyukovAidar) [inline test/visualTest.ts:841]: expect(links[0]...).toBeTrue();
const otherLinks = links.slice(1);
otherLinks.forEach((link)=>expect(link...).toBeFalse())

**Comment by MulyukovAidar** [comment]
LGTM

**Pull request merged**